File: src/main/java/com/github/windchopper/tools/process/browser/OperatingSystem.java
package com.github.windchopper.tools.process.browser;

import java.util.EnumSet;
import java.util.Optional;

public enum OperatingSystem {

    WINDOWS("win"),
    MACOS("mac"),
    LINUX("nux");

    private final String token;

    OperatingSystem(String token) {
        this.token = token;
    }

    public static Optional<OperatingSystem> detect() {
        Optional<String> operatingSystemNameProperty = Optional.of("os.name")
            .map(System::getProperty)
            .map(String::toLowerCase);

        return EnumSet.allOf(OperatingSystem.class).stream()
            .filter(system -> operatingSystemNameProperty
                .filter(name -> name.contains(system.token))
                .isPresent())
            .findFirst();
    }

}

File: src/main/java/com/github/windchopper/tools/process/browser/OperatingSystem.kt
@file:Suppress("unused")

package com.github.windchopper.tools.process.browser

import java.util.*

enum class OperatingSystem(private val token: String) {

    WIN32("win"), MACOS("mac"), LINUX("nux");

    companion object {

        fun detect(): OperatingSystem? {
            with (System.getProperty("os.name")?.toLowerCase()) {
                return EnumSet.allOf(OperatingSystem::class.java)
                    .firstOrNull { this
                        ?.contains(it.token)
                        ?:false }
            }
        }

    }

}